File: solutions/fortran/pangram/1/pangram.f90
module pangram
  implicit none
contains

  logical function is_pangram(sentence)
    character(*) :: sentence
    character(len=len(sentence)) :: s2
    integer :: i,j
    s2=""
    j=1
    if (sentence=="") then
        is_pangram=.false.
    else
      is_pangram=.true.
      do i=97,122
        is_pangram=is_pangram .and. (scan(sentence, char(i))/=0 .or. scan(sentence, char(i-32))/=0)
      end do 
    end if
   end function is_pangram

end module pangram

File: solutions/c/pangram/1/pangram.c
#include "pangram.h"

bool is_pangram(const char *sentence){
    bool answer = false;
    //test null case
    if (sentence != NULL) {
        if (strlen(sentence)>0){
            //replace non alphanumeric chatacters with ~ and lowercase alphanumeric
            unsigned int i=0;
            char *phrase=malloc(100*sizeof(char));
            strcpy(phrase,sentence);
            while (*(phrase+i)!='\0'){
                if (*(phrase+i)>=65 && *(phrase+i)<=90) *(phrase+i)=*(phrase+i)+32;
                else if (!(*(phrase+i)>=97 && *(phrase+i)<=122)) *(phrase+i)='~';
                i++;
            }
            //sort
            unsigned int j=0;
            char temp=0;
            while (j<strlen(phrase)){
                i=0;
                while (*(phrase+i)!='\0'){
                    if(*(phrase+i)>*(phrase+i+1) && *(phrase+i+1)!='\0') {
                        temp=*(phrase+i);
                        *(phrase+i)=*(phrase+i+1);
                        *(phrase+i+1)=temp;
                        }
                    i++;
                }
                j++;
            }
            //check if there are more than one alphanumeric character
            i=0;
            unsigned int count=0;
            while (*(phrase+i)!='\0'){
                if (*(phrase+i)!=*(phrase+i+1)) count++;
                i++;
            }
            if ((count==26) && *(phrase+i-1)=='z') answer=true;
            else if ((count==27) && *(phrase+i-1)=='~') answer=true;
        }
        
    }
    return answer;
}
